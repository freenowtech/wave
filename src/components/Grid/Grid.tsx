import React, { FC, PropsWithChildren } from 'react';
import styled from 'styled-components';
import { Spaces } from '../../essentials/Spaces/Spaces';
import { Box, BoxProps } from '../Box/Box';

const AMOUNT_OF_DECIMALS = 4;
const COLUMN_WIDTH = 100 / 12;
const GAP = Spaces[3];

type ColumnOffset = 0 | 1 | 2 | 3 | 4 | 5 | 6 | 7 | 8 | 9 | 10 | 11;
type ColumnSpan = 1 | 2 | 3 | 4 | 5 | 6 | 7 | 8 | 9 | 10 | 11 | 12;

const BorderBoxWrapper = styled(Box)`
    box-sizing: border-box;
`;

type RowProps = BoxProps;

const Row: FC<PropsWithChildren<RowProps>> = props => (
    <Box display="flex" flexWrap="wrap" marginRight={`-${GAP}`} {...props} />
);

interface ColumnProps extends BoxProps {
    span?: ColumnSpan;
    offset?: ColumnOffset;
}

const Column: FC<PropsWithChildren<ColumnProps>> = ({ span = 1, offset = 0, ...restProps }) => {
    const marginLeft = offset * COLUMN_WIDTH;
    const width = span * COLUMN_WIDTH;

    return (
        <BorderBoxWrapper
            flex={`0 0 ${width.toFixed(AMOUNT_OF_DECIMALS)}%`}
            marginLeft={marginLeft ? `${marginLeft.toFixed(AMOUNT_OF_DECIMALS)}%` : undefined}
            paddingRight={GAP}
            {...restProps}
        />
    );
};

export { Row, RowProps, Column, ColumnProps };
