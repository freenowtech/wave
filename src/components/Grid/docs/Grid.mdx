---
name: Grid System
menu: Components
route: /components/grid-system
---

import { Playground } from 'docz';
import { Button } from '../../Button/Button.tsx';
import { Headline } from '../../Headline/Headline.tsx';
import { Input } from '../../Input/Input.tsx';
import { Row, Column } from '../Grid.tsx';
import { StyledSystemLinks } from '../../../docs/StyledSystemLinks'
import { Text } from '../../Text/Text.tsx';
import { ColumnPropsTable } from './GridPropsTable';

# Grid System

The grid system is a set of components that can be used to create different layouts, composed of 12 vertical tracks of equal width.

## Properties

### Row

<StyledSystemLinks component="Row" supportedProps={ ['space', 'layout', 'position', 'color', 'flexbox', 'grid-layout', 'background'] }/>

### Column

<ColumnPropsTable /><br/>

<StyledSystemLinks component="Column" supportedProps={ ['space', 'layout', 'position', 'color', 'flexbox', 'grid-layout', 'background'] }/>

## Examples

The grid system is composed of 12 vertical tracks of equal width. Content can be placed in these tracks by using the `Row` and `Column` component.

<Playground>
<Row>
  <Column bg="#f1f2f4" textAlign="center">
    1
  </Column>
  <Column bg="#f1f2f4" textAlign="center">
    2
  </Column>
  <Column bg="#f1f2f4" textAlign="center">
    3
  </Column>
  <Column bg="#f1f2f4" textAlign="center">
    4
  </Column>
  <Column bg="#f1f2f4" textAlign="center">
    5
  </Column>
  <Column bg="#f1f2f4" textAlign="center">
    6
  </Column>
  <Column bg="#f1f2f4" textAlign="center">
    7
  </Column>
  <Column bg="#f1f2f4" textAlign="center">
    8
  </Column>
  <Column bg="#f1f2f4" textAlign="center">
    9
  </Column>
  <Column bg="#f1f2f4" textAlign="center">
    10
  </Column>
  <Column bg="#f1f2f4" textAlign="center">
    11
  </Column>
  <Column bg="#f1f2f4" textAlign="center">
    12
  </Column>
</Row>
</Playground>

`Column`s can occupy more than one vertical track, by changing the `span` prop to a number between 1 and 12.

<Playground>
<Row mb={3}>
  <Column bg="#f1f2f4" textAlign="center" span={2}>
    1
  </Column>
  <Column bg="#f1f2f4" textAlign="center" span={2}>
    2
  </Column>
  <Column bg="#f1f2f4" textAlign="center" span={2}>
    3
  </Column>
  <Column bg="#f1f2f4" textAlign="center" span={2}>
    4
  </Column>
  <Column bg="#f1f2f4" textAlign="center" span={2}>
    5
  </Column>
  <Column bg="#f1f2f4" textAlign="center" span={2}>
    6
  </Column>
</Row>
<Row mb={3}>
  <Column bg="#f1f2f4" textAlign="center" span={3}>
    7
  </Column>
  <Column bg="#f1f2f4" textAlign="center" span={3}>
    8
  </Column>
  <Column bg="#f1f2f4" textAlign="center" span={3}>
    9
  </Column>
  <Column bg="#f1f2f4" textAlign="center" span={3}>
    10
  </Column>
</Row>
<Row mb={3}>
  <Column bg="#f1f2f4" textAlign="center" span={4}>
    11
  </Column>
  <Column bg="#f1f2f4" textAlign="center" span={4}>
    12
  </Column>
  <Column bg="#f1f2f4" textAlign="center" span={4}>
    13
  </Column>
</Row>
<Row mb={3}>
  <Column bg="#f1f2f4" textAlign="center" span={6}>
    14
  </Column>
  <Column bg="#f1f2f4" textAlign="center" span={6}>
    15
  </Column>
</Row>
<Row>
  <Column bg="#f1f2f4" textAlign="center" span={12}>
    16
  </Column>
</Row>
</Playground>

`Column`s can be given an offset, by changing the `offset` prop to a number between 0 and 11.

<Playground>
  <Row>
    <Column bg="#f1f2f4" textAlign="center" span={2} offset={2}>
      1
    </Column>
    <Column bg="#f1f2f4" textAlign="center" span={2} offset={4}>
      2
    </Column>
  </Row>
</Playground>

The `Row` and `Column` components can be used to create different kinds of layouts, combining them with other components of the design system.

<Playground>
<Row mb={3}>
  <Column span={12}>
    <Headline as="h2">Create a free Business Account</Headline>
    <Text as="p">
      Make managing employee and guest travel easier. It only takes a couple
      of minutes to get started!
    </Text>
  </Column>
</Row>
<Row mb={3}>
  <Column span={6}>
    <Input label="First Name" width="100%" />
  </Column>
  <Column span={6}>
    <Input label="Last Name" width="100%" />
  </Column>
</Row>
<Row mb={3}>
  <Column span={6}>
    <Input label="Email" width="100%" />
  </Column>
  <Column span={2}>
    <Input label="Area Code" width="100%" />
  </Column>
  <Column span={4}>
    <Input label="Phone Number" width="100%" />
  </Column>
</Row>
<Row>
  <Column span={6} offset={6}>
    <Button width="100%">Sign up for FREE NOW for Business</Button>
  </Column>
</Row>
</Playground>
