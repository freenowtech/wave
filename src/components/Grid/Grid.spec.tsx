import { render } from '@testing-library/react';
import * as React from 'react';
import { Row, Column } from './Grid';

describe('Grid', () => {
    it('renders without props', () => {
        expect(
            render(
                <Row>
                    <Column>Column 1</Column>
                    <Column>Column 2</Column>
                    <Column>Column 3</Column>
                </Row>
            ).container.firstChild
        ).toMatchSnapshot();
    });

    it('renders columns with span', () => {
        expect(
            render(
                <Row>
                    <Column span={4}>Column 1</Column>
                    <Column span={8}>Column 2</Column>
                </Row>
            ).container.firstChild
        ).toMatchSnapshot();
    });

    it('renders columns with offset', () => {
        expect(
            render(
                <Row>
                    <Column>Column 1</Column>
                    <Column offset={10}>Column 2</Column>
                </Row>
            ).container.firstChild
        ).toMatchSnapshot();
    });
});
