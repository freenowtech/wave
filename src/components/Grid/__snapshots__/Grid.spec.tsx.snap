// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Grid renders columns with offset 1`] = `
.c0 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-wrap: wrap;
  -ms-flex-wrap: wrap;
  flex-wrap: wrap;
  margin-right: -1.5rem;
}

.c1 {
  -webkit-flex: 0 0 8.3333%;
  -ms-flex: 0 0 8.3333%;
  flex: 0 0 8.3333%;
  padding-right: 1.5rem;
}

.c3 {
  -webkit-flex: 0 0 8.3333%;
  -ms-flex: 0 0 8.3333%;
  flex: 0 0 8.3333%;
  margin-left: 83.3333%;
  padding-right: 1.5rem;
}

.c2 {
  box-sizing: border-box;
}

<div
  class="c0"
  display="flex"
>
  <div
    class="c1 c2"
  >
    Column 1
  </div>
  <div
    class="c3 c2"
  >
    Column 2
  </div>
</div>
`;

exports[`Grid renders columns with span 1`] = `
.c0 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-wrap: wrap;
  -ms-flex-wrap: wrap;
  flex-wrap: wrap;
  margin-right: -1.5rem;
}

.c1 {
  -webkit-flex: 0 0 33.3333%;
  -ms-flex: 0 0 33.3333%;
  flex: 0 0 33.3333%;
  padding-right: 1.5rem;
}

.c3 {
  -webkit-flex: 0 0 66.6667%;
  -ms-flex: 0 0 66.6667%;
  flex: 0 0 66.6667%;
  padding-right: 1.5rem;
}

.c2 {
  box-sizing: border-box;
}

<div
  class="c0"
  display="flex"
>
  <div
    class="c1 c2"
  >
    Column 1
  </div>
  <div
    class="c3 c2"
  >
    Column 2
  </div>
</div>
`;

exports[`Grid renders without props 1`] = `
.c0 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-wrap: wrap;
  -ms-flex-wrap: wrap;
  flex-wrap: wrap;
  margin-right: -1.5rem;
}

.c1 {
  -webkit-flex: 0 0 8.3333%;
  -ms-flex: 0 0 8.3333%;
  flex: 0 0 8.3333%;
  padding-right: 1.5rem;
}

.c2 {
  box-sizing: border-box;
}

<div
  class="c0"
  display="flex"
>
  <div
    class="c1 c2"
  >
    Column 1
  </div>
  <div
    class="c1 c2"
  >
    Column 2
  </div>
  <div
    class="c1 c2"
  >
    Column 3
  </div>
</div>
`;
